apps/emqx/src/emqx_connection.erl

```diff
@@ -1286,18 +1288,42 @@ set_tcp_keepalive({Type, Id}) ->
 
 -spec graceful_shutdown_transport(atom(), state()) -> state().
 graceful_shutdown_transport(
-    kicked,
+    Reason,
     S = #state{
         transport = emqx_quic_stream,
         socket = Socket
     }
-) ->
-    _ = emqx_quic_stream:shutdown(Socket, read_write, 1000),
-    S#state{sockstate = closed};
+) when Reason =:= takenover; Reason =:= kicked; Reason =:= discarded ->
+    _ = emqx_quic_stream:abort_read(Socket, Reason),
+    S#state{sockstate = read_aborted};
 graceful_shutdown_transport(_Reason, S = #state{transport = Transport, socket = Socket}) ->
     _ = Transport:shutdown(Socket, read_write),
     S#state{sockstate = closed}.
 
+%% @doc wait for stream close or stream read shutdown complete
+%%      this also ensures write shutdown are completed.
+%%      see emqx_quic_stream:abort_read/2
+%% @end
+-spec wait_for_quic_stream_close(state()) -> state().
+wait_for_quic_stream_close(
+    State = #state{
+        transport = emqx_quic_stream,
+        socket = {quic, _Conn, Stream, _},
+        sockstate = read_aborted
+    }
+) ->
+    receive
+        %% We are expecting peer to close the stream
+        {quic, Evtname, Stream, _} when
+            Evtname =:= peer_send_shutdown orelse
+                Evtname =:= peer_send_aborted orelse
+                Evtname =:= stream_closed
+        ->
+            ok
+    after 3000 ->
+        ok
```

Comment:
Nit. It's a bit unobvious timeout value. Why 3 seconds? Maybe macrodef or at least a comment describing how it was chosen?